pkg/controller/workload/jobframework/workload_names.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jobframework
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	hashLength = 5
+	// 253 is the maximal length for a CRD name. We need to subtract one for '-', and the hash length.
+	maxPrefixLength = 252 - hashLength
+)
+
+func GetWorkloadNameForOwner(owner *metav1.OwnerReference) string {
+	prefixedName := strings.ToLower(owner.Kind) + "-" + owner.Name
+	if len(prefixedName) > maxPrefixLength {
+		prefixedName = prefixedName[:maxPrefixLength]
+	}
+	return prefixedName + "-" + getHash(owner)[:hashLength]
+}
+
+func getHash(owner *metav1.OwnerReference) string {
+	h := sha1.New()
+	h.Write([]byte(owner.Kind))
+	h.Write([]byte("\n"))
+	h.Write([]byte(owner.APIVersion))
```

Comment:
I wonder if we should exclude the version. In the case of Job, we could just add `batch`, not `batch/v1`.
Does it matter?
On one hand the reconcilers don't care about the name, they just check that there is one workload.
Only the job webhook cares. Currently the job webhook is in the same binary as the mpi-operator, so they will always change together. But later they will be in different binaries so they might be using different API versions. Also we can have a similar problem when there is leader election. A webhook can be in an older version than the leader reconciler.

Reply:
I see, it means unnecessary re-creation and rescheduling of MPIJobs when the version is bumped from v2beta1 to v2beta2 to v2. 

I now parse the GV from the OwnerReference and don't use the Version in the hash.